### PR TITLE
Eliminate some interface dispatch around object pool use

### PR DESCRIPTION
**Pull request by geeknoid**
- Use concrete types for fields instead of interface types, which avoids interface dispatch and enables inlining.

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

File: src/Polly.Core/Utils/CancellationTokenSourcePool.cs [hunk 1]
```diff
@@ -1,33 +1,78 @@
...
+
+        return source;
     }
 
-    protected abstract CancellationTokenSource GetCore(TimeSpan delay);
+    [SuppressMessage("Performance", "CA1822:Mark members as static", Justification = "Can't be static on >= .NET 6")]
+    public void Return(CancellationTokenSource source)
+    {
+#if NET6_0_OR_GREATER
+        if (this == SystemInstance && source.TryReset())
```
Comment (martintmk) [inline src/Polly.Core/Utils/CancellationTokenSourcePool.cs:67]: nit: Isn't now this extra if check that's always executed worse than the method call?

**Review by martintmk — COMMENTED** [review]
I am wondering what's the real impact of this change.
Can you run the `Polly.Core.Benchmarks` (at least for timeout startegy) to see the difference?

File: src/Polly.Core/Utils/CancellationTokenSourcePool.cs (hunk 1 above)
Comment (geeknoid) [reply src/Polly.Core/Utils/CancellationTokenSourcePool.cs:67]: Yes, there is an extra "if" check when on .NET 6 or above. But you trade this for eliminating virtual dispatch and enabling inlining.

**Comment by geeknoid** [comment]
@martintmk Here are the before and after numbers:

BEFORE

```
|                                                           Method |       Job |                                         Arguments | NuGetReferences |            Mean |        Error |       StdDev | Ratio | RatioSD |   Gen0 | Allocated | Alloc Ratio |
|----------------------------------------------------------------- |---------- |-------------------------------------------------- |---------------- |----------------:|-------------:|-------------:|------:|--------:|-------:|----------:|------------:|
|                                     Timeout_Synchronous_Succeeds |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|                                     Timeout_Synchronous_Succeeds | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        186.9 ns |      0.83 ns |      0.77 ns |     ? |       ? | 0.0610 |     640 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
|                                    Timeout_Asynchronous_Succeeds |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|                                    Timeout_Asynchronous_Succeeds | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        234.0 ns |      1.36 ns |      1.27 ns |     ? |       ? | 0.0718 |     752 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
|             Timeout_Asynchronous_Succeeds_With_CancellationToken |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|             Timeout_Asynchronous_Succeeds_With_CancellationToken | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        242.7 ns |      3.69 ns |      3.08 ns |     ? |       ? | 0.0715 |     752 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
|                         Timeout_Synchronous_With_Result_Succeeds |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|                         Timeout_Synchronous_With_Result_Succeeds | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        182.0 ns |      1.94 ns |      1.81 ns |     ? |       ? | 0.0527 |     552 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
|                        Timeout_Asynchronous_With_Result_Succeeds |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|                        Timeout_Asynchronous_With_Result_Succeeds | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        209.7 ns |      1.74 ns |      1.55 ns |     ? |       ? | 0.0489 |     512 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
| Timeout_Asynchronous_With_Result_Succeeds_With_CancellationToken |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
| Timeout_Asynchronous_With_Result_Succeeds_With_CancellationToken | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        210.4 ns |      2.28 ns |      2.13 ns |     ? |       ? | 0.0489 |     512 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
|                        Timeout_Asynchronous_Times_Out_Optimistic |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|                        Timeout_Asynchronous_Times_Out_Optimistic | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default | 15,779,961.2 ns | 71,062.50 ns | 66,471.90 ns |     ? |       ? |      - |    1369 B |           ? |
```

AFTER

```
|                                                           Method |       Job |                                         Arguments | NuGetReferences |            Mean |        Error |       StdDev | Ratio | RatioSD |   Gen0 | Allocated | Alloc Ratio |
|----------------------------------------------------------------- |---------- |-------------------------------------------------- |---------------- |----------------:|-------------:|-------------:|------:|--------:|-------:|----------:|------------:|
|                                     Timeout_Synchronous_Succeeds |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|                                     Timeout_Synchronous_Succeeds | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        181.2 ns |      1.83 ns |      1.71 ns |     ? |       ? | 0.0610 |     640 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
|                                    Timeout_Asynchronous_Succeeds |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|                                    Timeout_Asynchronous_Succeeds | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        234.8 ns |      1.35 ns |      1.27 ns |     ? |       ? | 0.0718 |     752 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
|             Timeout_Asynchronous_Succeeds_With_CancellationToken |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|             Timeout_Asynchronous_Succeeds_With_CancellationToken | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        245.7 ns |      2.33 ns |      1.94 ns |     ? |       ? | 0.0715 |     752 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
|                         Timeout_Synchronous_With_Result_Succeeds |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|                         Timeout_Synchronous_With_Result_Succeeds | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        172.7 ns |      0.59 ns |      0.52 ns |     ? |       ? | 0.0527 |     552 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
|                        Timeout_Asynchronous_With_Result_Succeeds |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|                        Timeout_Asynchronous_With_Result_Succeeds | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        204.0 ns |      0.97 ns |      0.81 ns |     ? |       ? | 0.0489 |     512 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
| Timeout_Asynchronous_With_Result_Succeeds_With_CancellationToken |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
| Timeout_Asynchronous_With_Result_Succeeds_With_CancellationToken | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default |        203.9 ns |      0.75 ns |      0.58 ns |     ? |       ? | 0.0489 |     512 B |           ? |
|                                                                  |           |                                                   |                 |                 |              |              |       |         |        |           |             |
|                        Timeout_Asynchronous_Times_Out_Optimistic |     Polly |  /p:BenchmarkFromNuGet=True,/p:SignAssembly=false |     Polly 7.2.3 |              NA |           NA |           NA |     ? |       ? |      - |         - |           ? |
|                        Timeout_Asynchronous_Times_Out_Optimistic | Polly-dev | /p:BenchmarkFromNuGet=False,/p:SignAssembly=false |         Default | 15,772,982.1 ns | 54,245.85 ns | 50,741.60 ns |     ? |       ? |      - |    1369 B |           ? |
```

**Review by martincostello — APPROVED** [review]

**Comment by geeknoid** [comment]
The benchmark around combining the pool types into one were inconclusive, so I just undid that part of the PR.

**Review by martintmk — APPROVED** [review]

**Pull request merged**